src/WebJobs.Script/Eventing/File/FileWatcherEventSource.cs

```diff
@@ -27,12 +28,22 @@ public sealed class FileWatcherEventSource : IDisposable
             _eventManager = eventManager;
             _fileWatcher = new AutoRecoveringFileSystemWatcher(path, filter, includeSubdirectories, changeTypes, traceWriter);
             _fileWatcher.Changed += FileChanged;
+            _traceWriter = traceWriter;
         }
 
         private void FileChanged(object sender, FileSystemEventArgs e)
         {
-            var fileEvent = new FileEvent(_source, e);
-            _eventManager.Publish(fileEvent);
+            // This handler is called on a background thread, so any exceptions thrown will crash
+            // the process. Handle and log all errors instead.
+            try
+            {
+                var fileEvent = new FileEvent(_source, e);
+                _eventManager.Publish(fileEvent);
+            }
+            catch (Exception ex)
+            {
+                _traceWriter?.Error($"Error handling '{e.ChangeType}' notification for '{e.FullPath}'.", ex);
```

Comment:
Note that this is a behavior change as now the process won't restart on an exception here. This could lead to rare unintended side effects. If we still want that behavior, I could perform the restart in more graceful way by calling `IScriptHostEnvironment.Shutdown()` after logging the error. Thoughts?